File: VK_Client/app/src/main/java/com/nethergrim/vk/models/PendingMessage.java
package com.nethergrim.vk.models;

import android.os.Parcel;
import android.os.Parcelable;
import android.support.annotation.NonNull;

import java.util.List;
import java.util.Random;

import io.realm.RealmList;

/**
 * Created on 21.08.16.
 */

public class PendingMessage implements Parcelable {

    public static final Creator<PendingMessage> CREATOR = new Creator<PendingMessage>() {
        @Override
        public PendingMessage createFromParcel(Parcel in) {
            return new PendingMessage(in);
        }

        @Override
        public PendingMessage[] newArray(int size) {
            return new PendingMessage[size];
        }
    };
    private static final Random rand = new Random();
    private long randomId;
    private String text;
    private String forwardedMessageIds;
    private StickerLocal sticker;
    private long peerId;

    public PendingMessage(long randomId, String text, String forwarded, long stickerId, long peerId) {
        this.randomId = randomId;
        this.text = text;
        this.forwardedMessageIds = forwarded;
        this.peerId = peerId;
        if (stickerId > 0) {
            sticker = new StickerLocal(stickerId, null);
        }
    }

    public PendingMessage() {
    }

    protected PendingMessage(Parcel in) {
        randomId = in.readLong();
        text = in.readString();
        forwardedMessageIds = in.readString();
        sticker = in.readParcelable(StickerLocal.class.getClassLoader());
        peerId = in.readLong();
    }

    public static long generateRandomId() {
        return rand.nextLong();
    }

    public static PendingMessage fromMessage(@NonNull Message message) {
        PendingMessage pendingMessageJava = new PendingMessage();
        pendingMessageJava.randomId = message.getId();
        pendingMessageJava.text = message.getBody();
        pendingMessageJava.peerId = message.getPeerId();
        pendingMessageJava.forwardedMessageIds = null; // TODO handle forwaded messages from here
        RealmList<Attachment> attachments = message.getAttachments();
        for (int i = 0, attachmentsSize = attachments.size(); i < attachmentsSize; i++) {
            Attachment attachment = attachments.get(i);
            if (attachment.getSticker() != null) {
                Sticker sticker = attachment.getSticker();
                pendingMessageJava.sticker = new StickerLocal(Long.parseLong(sticker.getId()), sticker.getPhoto256());
            }
        }
        return pendingMessageJava;
    }

    public static List<PendingMessage> fromMessages(List<Message> result) {
        return rx.Observable.from(result).map(PendingMessage::fromMessage).toList().toBlocking().first();
    }

    public long getRandomId() {
        return randomId;
    }

    public void setRandomId(long randomId) {
        this.randomId = randomId;
    }

    public String getText() {
        return text;
    }

    public void setText(String text) {
        this.text = text;
    }

    public String getForwardedMessageIds() {
        return forwardedMessageIds;
    }

    public void setForwardedMessageIds(String forwardedMessageIds) {
        this.forwardedMessageIds = forwardedMessageIds;
    }

    public StickerLocal getSticker() {
        return sticker;
    }

    public void setSticker(StickerLocal sticker) {
        this.sticker = sticker;
    }

    public long getPeerId() {
        return peerId;
    }

    public void setPeerId(long peerId) {
        this.peerId = peerId;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel parcel, int i) {
        parcel.writeLong(randomId);
        parcel.writeString(text);
        parcel.writeString(forwardedMessageIds);
        parcel.writeParcelable(sticker, i);
        parcel.writeLong(peerId);
    }

    public Message mapToMessage(long currentUserId) {
        // TODO implement
        return null;
    }
}

File: VK_Client/app/src/main/java/com/nethergrim/vk/models/PendingMessage.kt
@file:Suppress("NOTHING_TO_INLINE")

package com.nethergrim.vk.models

import android.os.Parcel
import android.os.Parcelable
import com.nethergrim.vk.utils.ConversationUtils
import java.util.*


data class PendingMessage(val randomId: Long, val text: String?, val forwardedMessageIds: String?, val stickerId: Long?, val peerId: Long) : Parcelable {

    constructor(source: Parcel) : this(source.readLong(), source.readString(), source.readString(), source.readLong(), source.readLong())

    override fun describeContents() = 0

    override fun writeToParcel(dest: Parcel?, flags: Int) {
        dest?.writeLong(randomId)
        dest?.writeString(text)
        dest?.writeString(forwardedMessageIds)
        dest?.writeLong(stickerId ?: 0L)
        dest?.writeLong(peerId)
    }

    fun mapToMessage(currentUserId: Long): Message {
        val message = Message()
        message.id = this.randomId
        message.read_state = 0
        message.out = 1
        message.isPending = true
        message.date = System.currentTimeMillis() / 1000
        message.from_id = currentUserId
        message.body = this.text
        if (ConversationUtils.isPeerIdAGroupChat(this.peerId)) {
            message.chat_id = ConversationUtils.getConversationIdFromPeerId(this.peerId)
        } else {
            message.user_id = ConversationUtils.getUserIdFromPeerId(this.peerId)
        }
        // TODO #134 add support for stickers and forwarded messages
        return message
    }

    companion object {

        val rand = Random()

        @Suppress("unused")
        @JvmField val CREATOR: Parcelable.Creator<PendingMessage> = object : Parcelable.Creator<PendingMessage> {
            override fun createFromParcel(source: Parcel): PendingMessage = PendingMessage(source)
            override fun newArray(size: Int): Array<PendingMessage?> = arrayOfNulls(size)
        }

        inline fun generateRandomId(): Long {
            return rand.nextLong()
        }

        inline fun fromMessage(message: Message): PendingMessage {
            // TODO #134 add support for stickers and forwarded messages
            val pendingMessage = PendingMessage(message.id, message.body, null, null, message.peerId)

            return pendingMessage
        }

        inline fun fromMessages(messages: List<Message>): List<PendingMessage> {
            return messages.mapNotNull { fromMessage(it) }
        }
    }
}